File: js-sdk/qued.ts
import { createClient, type RedisClientType } from "redis";
import { v4 as uuidv4 } from "uuid";

export interface QuedOptions {
  name: string;
  redisUrl: string;
  maxTries?: number;
}

export interface QuedMessage<T = any> {
  id: string;
  type?: string;
  payload: T;
  created_at: string;
  attempts: number;
}

export class Qued {
  private name: string;
  private deadName: string;
  private maxTries: number;
  private client: RedisClientType;

  constructor({ name, redisUrl, maxTries = 3 }: QuedOptions) {
    this.name = name;
    this.deadName = `${name}:dead`;
    this.maxTries = maxTries;
    this.client = createClient({ url: redisUrl });
    this.client.connect();
  }

  /**
   * Enqueue a message to the queue.
   * @param type Message type (optional)
   * @param payload Message payload
   * @returns The message ID
   */
  async enqueue<T = any>(
    type: string | undefined,
    payload: T
  ): Promise<string> {
    const msg: QuedMessage<T> = {
      id: uuidv4(),
      type,
      payload,
      created_at: new Date().toISOString(),
      attempts: 0,
    };

    await this.client.lPush(this.name, JSON.stringify(msg));
    return msg.id;
  }

  /**
   * Dequeue a message from the queue (blocking pop).
   * @param timeout Timeout in seconds (0 = block forever)
   * @returns The message or null if timeout
   */
  async dequeue<T = any>(timeout = 0): Promise<QuedMessage<T> | null> {
    // BRPOP returns { key, element } or null
    const result = await this.client.brPop(this.name, timeout);
    if (!result) return null;
    return JSON.parse(result.element) as QuedMessage<T>;
  }

  /**
   * Move a message to the dead-letter queue.
   * @param msg The message to move
   */
  async fail<T = any>(msg: QuedMessage<T>): Promise<void> {
    msg.attempts = (msg.attempts || 0) + 1;
    await this.client.lPush(this.deadName, JSON.stringify(msg));
  }

  /**
   * Retry a message, or move to dead-letter if maxTries exceeded.
   * @param msg The message to retry
   */
  async retry<T = any>(msg: QuedMessage<T>): Promise<void> {
    msg.attempts = (msg.attempts || 0) + 1;
    if (msg.attempts >= this.maxTries) {
      await this.fail(msg);
    } else {
      await this.client.lPush(this.name, JSON.stringify(msg));
    }
  }

  /**
   * Dequeue a message from the dead-letter queue (blocking pop).
   * @param timeout Timeout in seconds (0 = block forever)
   * @returns The message or null if timeout
   */
  async deadLetter<T = any>(timeout = 0): Promise<QuedMessage<T> | null> {
    const result = await this.client.brPop(this.deadName, timeout);
    if (!result) return null;
    return JSON.parse(result.element) as QuedMessage<T>;
  }

  /**
   * Disconnect from Redis.
   */
  async disconnect(): Promise<void> {
    await this.client.close();
  }
}

File: go-sdk/qued.go
package qued

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"time"

	"github.com/google/uuid"
	"github.com/redis/go-redis/v9"
)

// Message is the structure for queued items.
type Message struct {
	ID        string `json:"id"`
	Type      string `json:"type,omitempty"`
	Payload   any    `json:"payload"`
	CreatedAt string `json:"created_at"`
	Attempts  int    `json:"attempts,omitempty"`
}

// Queue wraps Redis connection and queue name.
type Queue struct {
	client   *redis.Client
	name     string
	deadName string
	maxTries int
}

// NewQueue creates a new Queue instance.
func NewQueue(name, redisURL string, maxTries int) *Queue {
	opts, err := redis.ParseURL(redisURL)
	if err != nil {
		panic(fmt.Errorf("invalid Redis URL: %w", err))
	}
	return &Queue{
		client:   redis.NewClient(opts),
		name:     name,
		deadName: fmt.Sprintf("%s:dead", name),
		maxTries: maxTries,
	}
}

// Enqueue ads a message to the queue.
func (q *Queue) Enqueue(ctx context.Context, msgType string, payload any) (string, error) {
	msg := Message{
		ID:        uuid.New().String(),
		Type:      msgType,
		Payload:   payload,
		CreatedAt: time.Now().UTC().Format(time.RFC3339),
		Attempts:  0,
	}

	data, err := json.Marshal(msg)
	if err != nil {
		return "", err
	}

	if err := q.client.LPush(ctx, q.name, data).Err(); err != nil {
		return "", err
	}

	return msg.ID, nil
}

// Dequeue retrieves a message from the queue (blocking).
func (q *Queue) Dequeue(ctx context.Context, timeout time.Duration) (*Message, error) {
	res, err := q.client.BRPop(ctx, timeout, q.name).Result()
	if err != nil {
		if errors.Is(err, redis.Nil) {
			return nil, nil
		}
		return nil, err
	}

	var msg Message
	if err := json.Unmarshal([]byte(res[1]), &msg); err != nil {
		return nil, err
	}

	return &msg, nil
}

// Fail moves a message to the dead-letter queue.
func (q *Queue) Fail(ctx context.Context, msg *Message) error {
	msg.Attempts++
	data, err := json.Marshal(msg)

	if err != nil {
		return err
	}

	return q.client.LPush(ctx, q.deadName, data).Err()
}

// Retry re-enqueues a message, or moves to dead-letter if maxTries exceeded.
func (q *Queue) Retry(ctx context.Context, msg *Message) error {
	msg.Attempts++
	if msg.Attempts >= q.maxTries {
		return q.Fail(ctx, msg)
	}
	data, err := json.Marshal(msg)
	if err != nil {
		return err
	}
	return q.client.LPush(ctx, q.name, data).Err()
}

// DeadLetter retrieves a message from the dead-letter queue (blocking).
func (q *Queue) DeadLetter(ctx context.Context, timeout time.Duration) (*Message, error) {
	res, err := q.client.BRPop(ctx, timeout, q.deadName).Result()
	if err != nil {
		if errors.Is(err, redis.Nil) {
			return nil, nil
		}
		return nil, err
	}

	var msg Message
	if err := json.Unmarshal([]byte(res[1]), &msg); err != nil {
		return nil, err
	}

	return &msg, nil
}

// Close the queue and release resources.
func (q *Queue) Close() error {
	return q.client.Close()
}
